lib/Support/Check.cpp

```diff
@@ -0,0 +1,81 @@
+//===- Check.cpp - Check test results  ------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//
+//===----------------------------------------------------------------------===//
+
+#include "Support/Check.h"
+
+static bool isDenorm(float F) {
+  return (std::numeric_limits<float>::denorm_min() <= F &&
+          F < std::numeric_limits<float>::min()) ||
+         (-std::numeric_limits<float>::min() < F &&
+          F <= -std::numeric_limits<float>::denorm_min());
+}
+
+static bool compareFloatULP(const float &FSrc, const float &FRef,
+                            unsigned ULPTolerance, offloadtest::DenormMode DM) {
+  if (FSrc == FRef)
+    return true;
+  if (std::isnan(FSrc))
+    return std::isnan(FRef);
+  if (DM == offloadtest::DenormMode::Any) {
+    // If denorm expected, output can be sign preserved zero. Otherwise output
+    // should pass the regular ulp testing.
+    if (isDenorm(FRef) && FSrc == 0 && std::signbit(FSrc) == std::signbit(FRef))
+      return true;
+  }
+  // For FTZ or Preserve mode, we should get the expected number within
+  // ULPTolerance for any operations.
+  int Diff = *((const uint32_t *)&FSrc) - *((const uint32_t *)&FRef);
+  unsigned int AbsDiff = Diff < 0 ? -Diff : Diff;
+  return AbsDiff <= ULPTolerance;
+}
+
+static bool testBufferExact(offloadtest::Buffer *B1, offloadtest::Buffer *B2) {
+  if (B1->size() != B2->size())
+    return false;
+  for (uint32_t I = 0; I < B1->size(); ++I) {
+    if (B1->Data[I] != B2->Data[I])
+      return false;
+  }
+  return true;
+}
+
+static bool testBufferFuzzy(offloadtest::Buffer *B1, offloadtest::Buffer *B2,
+                            unsigned ULPT, offloadtest::DenormMode DM) {
+  switch (B1->Format) {
+  case offloadtest::DataFormat::Float32: {
+    if (B1->Size != B2->Size)
+      return false;
+    llvm::MutableArrayRef<float> Arr1(reinterpret_cast<float *>(B1->Data.get()),
+                                      B1->Size / sizeof(float));
+    assert(B2->Format == offloadtest::DataFormat::Float32 &&
+           "Buffer types must be the same");
+    llvm::MutableArrayRef<float> Arr2(reinterpret_cast<float *>(B2->Data.get()),
+                                      B2->Size / sizeof(float));
+    for (unsigned I = 0; I < Arr1.size(); ++I) {
+      if (!compareFloatULP(Arr1[I], Arr2[I], ULPT, DM))
+        return false;
+    }
+    return true;
+  }
+  default:
+    llvm_unreachable("Only float types are supported by the fuzzy test.");
+  }
+  return false;
+}
+
+bool getResult(offloadtest::Result R) {
```

Comment:
Instead of returning bool, if these functions returned `llvm::Error` we could embed semantic information about the failure into the error. That could allow for more informative results. For example in the ULP compare you could capture the actual and expected values.

Also `llvm::Error` has this neat feature `joinError`, which allows you to build arrays of errors, so you can report back not just the first failure but a list of failures.

Even if we don't go to that full complexity in the initial implementation it is probably worth using `llvm::Error` to report success/failure just to have the ability to improve the diagnostics in the future as we extend this to support alternate verification rules.

Reply:
Ok I've replaced this function's usage of bool with llvm:Error.